### DISCUSSION TITLE
Problem with CSP for Ajax-Requests

### DISCUSSION THREAD

**Discussion by pneumcke** [primefaces/community#2432] [PrimeFaces] [answered]
Since 14.0.2 probably caused by https://github.com/primefaces/primefaces/commit/2b0b11e490f2110a1170db6cd056d9343a4c7ceb all Ajax requests fail in our application due to CSP violations.

The previous implementation in CSPState had two ways of accessing the nonce:
  this.nonce = (String)viewMap.get("primefaces.nonce");
and
  this.nonce = (String)this.context.getExternalContext().getRequestParameterMap().get("primefaces.nonce");

For the Ajax requests, the first access returns "null" - which was OK since the nonce was later accessed via the request parameter. 

Since 14.0.2 this behaviour does not work anymore. The view root exists and has 16 parameters (WELD%ManagedBean, omnifaces.Socket, ...) - but no "primefaces.nonce".

The JSF implementation is myfaces 4.0.1

Do you have any ideas? Are we doing something wrong?

**Comment by melloware** [comment]
@pneumcke this new security feature was added to CSP: https://github.com/primefaces/primefaces/issues/12140

I am using it in production without issue so can you put together a [PrimeFaces Test](https://github.com/primefaces/primefaces-test) reproducer showing your issue?

**Reply by tandraschko** [reply]
@melloware did you test this feature with stateless views?

**Reply by melloware** [reply]
No that is a good question let me test with a Stateless view or add an Integration test.  Our full test suite runs with CSP enabled.

**Reply by melloware** [reply]
it looks like we have autoUpdate002.xhtml already set to `transient="true"` so we have at least one stateless view integration test with CSP.

https://github.com/primefaces/primefaces/blob/master/primefaces-integration-tests/src/main/webapp/autoupdate/autoUpdate002.xhtml#L8

**Reply by tandraschko** [reply, marked as answer]
hmm
maybe we should add a "!= null" check here: https://github.com/primefaces/primefaces/blob/master/primefaces/src/main/java/org/primefaces/csp/CspState.java#L59
and also init nonceViewState  with null, no need a empty string here

Then we are more safe

**Reply by tandraschko** [reply]
but actually we would need a sample first as we would propably workaround a real bug

**Reply by melloware** [reply]
We have one situation where this is happening in one of our 70 apps @tandraschko but @FlipwartHog and I cannot reproduce it.  But our app using an `<h:commandLink ajax="false target="_blank" />` if that helps at all.  I am sure it doesn't.

i tried to create a standalone reproducer but i must be missing something that I can't reproduce.  I think the issue is absolutely as you suspected right here `if (context.isPostback() && !Objects.equals(nonceViewState, nonceRequest)) {` in our scenario.

Just because they are not equal it doesn't `NULL != ""`. I am going to convert this to an issue

**Reply by melloware** [reply]
Issue and PR created: https://github.com/primefaces/primefaces/issues/12399

**Reply by melloware** [reply]
@pneumcke please try 14.0.4

**Reply by pneumcke** [reply]
14.0.4 solves my issue with the CSP check.
Thanks for your fast response!
